### PR TITLE
[ownership] Update LowerAggregateInstrs for Ownership

### PR DESCRIPTION
**Pull request by gottesmm**
There are two different PRs here:

1. There are parts of lower aggregate inserts where I need to use emit{Load,Store}ValueOperation to emit code in both ossa/non-ossa mode but for which in non-ossa mode, we need to use the emitLowered* APIs. I exposed these on SILBuilder so we can keep the old codegen in non-ossa mode.

2. In the second commit I update for OSSA. The main difference is that we do not explode anything today: instead we just promote loadable typed copy_addr, destroy_addr to value operations. The reason why I am doing this is that originally the exploding was done to make it easier for the low level ARC optimizer to optimize (since aggregates with multiple-nontrivial mess with its usage of the concept of "RC Identity"). The high level ARC optimizer doesn't care about this and in fact it is just noise that gets in the way of optimizing and makes its job harder. Additionally since this is a function pass, after we lower ownership we will re-run lower aggregate instrs ensuring that the low level arc optimizer still sees what it is looking for.

### CODE REVIEW

**Comment by gottesmm** [comment]
NOTE: This doesn't move ownership lowering. This just is going to add the tests (to lock down where we are) and make sure that we do not change anything as a result of this change when running in non-ossa mode.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
The code-size aspect is interesting since that suggests a real difference in rr or something.

**Review by atrick — APPROVED** [review]
Very nice!

File: lib/SIL/IR/TypeLowering.cpp [hunk 1]
```diff
@@ -819,16 +900,12 @@ namespace {
     void forEachNonTrivialChild(SILBuilder &B, SILLocation loc,
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/IR/TypeLowering.cpp:900]: I don't know why either.

File: lib/SIL/IR/TypeLowering.cpp [hunk 2]
```diff
@@ -761,6 +782,40 @@ namespace {
...
+
+      SILValue loadValue =
+          B.createLoad(loc, addr, LoadOwnershipQualifier::Unqualified);
+
+      // If we do not have a copy, just return the value...
+      if (qual != LoadOwnershipQualifier::Copy)
+        return loadValue;
+
+      // Otherwise, emit the copy value operation.
+      return B.emitLoweredCopyValueOperation(loc, loadValue, expansionKind);
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/IR/TypeLowering.cpp:800]: It is handled automatically by the API.

[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/IR/TypeLowering.cpp:800]: Oh ok. Misread your msg. If we have a load-take, we just return the unqualified loaded value.

File: lib/SIL/IR/TypeLowering.cpp [hunk 3]
```diff
@@ -761,6 +782,40 @@ namespace {
...
+        B.createStore(loc, value, addr, qual);
+        return;
+      }
+
+      if (qual == StoreOwnershipQualifier::Assign) {
+        SILValue oldValue = B.emitLoadValueOperation(
+            loc, addr, LoadOwnershipQualifier::Unqualified);
+        B.emitLoweredDestroyValueOperation(loc, oldValue, expansionKind);
+      }
+      B.createStore(loc, value, addr, StoreOwnershipQualifier::Unqualified);
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/IR/TypeLowering.cpp:817]: I can do that.

File: lib/SIL/IR/TypeLowering.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SIL/IR/TypeLowering.cpp:800]: I think I misread the code. Anyway, a covered switch over the qualifiers just helps with the mental process of ruling out all cases.

**Comment by gottesmm** [comment]
Just did a rebase. I am curious about the code-size difference. I am going to look at that.

**Comment by gottesmm** [comment]
The reason why I think it is interesting is that I have not moved ownership past lower aggregate instrs, so we are really only testing that my refactoring doesn't change the output in any way when we run lower aggregate intrs on non-ossa code. So I would expect this to be NFCish.

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci build toolchain

**Comment by gottesmm** [comment]
@airspeedswift I was right that it was interesting behavior and you were right that it had to do with inlining! We are for some reason not inlining:

  // function_ref specialized UnsafeMutableBufferPointer.subscript.read                                                                  
  %16 = function_ref @$sSryxSicir9Phonebook6RecordV_Tg5 : $@yield_once @convention(method) (Int, UnsafeMutableBufferPointer<Record>) -> \
@yields @in_guaranteed Record // users: %36, %17 

Record is a String. Since we have a coroutine here and we have tight guaranteed scopes, we immediately retain in the caller before the coroutine ends. The end result is that some retain/releases of Strings can not be eliminated anymore. I imagine this isn't in fast code.

That being said:

1. I imagine that UnsafeMutableBufferPointer should be always inline. What do you think about doing that.
2. @eeckstein It may be worth for the inliner to look for these types of return values from coroutines and bump the inline threshold so that we can eliminate ARC traffic. We can't move the retain past the end_apply since we have tight scopes = (.

**Comment by eeckstein** [comment]
@gottesmm The inliner should always inline this function (it's very small). Can you create an isolated test case and send it over to me?

**Comment by gottesmm** [comment]
@atrick @eeckstein it could be due to pass manager not rerunning correctly.

**Comment by gottesmm** [comment]
Ok. Here is the only difference before/after. I think that it shows that we are missing some simplification peepholes. Consider:

```
  %37 = load %34 : $*Record                       // users: %57, %47, %38
  %38 = struct_extract %37 : $Record, #Record.first // user: %39
  %39 = struct_extract %38 : $String, #String._guts // user: %40
  %40 = struct_extract %39 : $_StringGuts, #_StringGuts._object // users: %42, %41
  %41 = struct_extract %40 : $_StringObject, #_StringObject._countAndFlagsBits // user: %44
  %42 = struct_extract %40 : $_StringObject, #_StringObject._object // users: %44, %43
  strong_retain %42 : $Builtin.BridgeObject       // id: %43
  %44 = struct $_StringObject (%41 : $UInt64, %42 : $Builtin.BridgeObject) // user: %45
  %45 = struct $_StringGuts (%44 : $_StringObject) // user: %46
  %46 = struct $String (%45 : $_StringGuts)       // user: %56
  %47 = struct_extract %37 : $Record, #Record.last // user: %48
  %48 = struct_extract %47 : $String, #String._guts // user: %49
  %49 = struct_extract %48 : $_StringGuts, #_StringGuts._object // users: %51, %50
  %50 = struct_extract %49 : $_StringObject, #_StringObject._countAndFlagsBits // user: %53
  %51 = struct_extract %49 : $_StringObject, #_StringObject._object // users: %53, %52
  strong_retain %51 : $Builtin.BridgeObject       // id: %52
  %53 = struct $_StringObject (%50 : $UInt64, %51 : $Builtin.BridgeObject) // user: %54
  %54 = struct $_StringGuts (%53 : $_StringObject) // user: %55
  %55 = struct $String (%54 : $_StringGuts)       // user: %56
  %56 = struct $Record (%46 : $String, %55 : $String)
  store %37 to %36 : $*Record                     // id: %57
```

vs

```
  %37 = load %34 : $*Record                       // users: %47, %38
  %38 = struct_extract %37 : $Record, #Record.first // user: %39
  %39 = struct_extract %38 : $String, #String._guts // user: %40
  %40 = struct_extract %39 : $_StringGuts, #_StringGuts._object // users: %42, %41
  %41 = struct_extract %40 : $_StringObject, #_StringObject._countAndFlagsBits // user: %44
  %42 = struct_extract %40 : $_StringObject, #_StringObject._object // users: %44, %43
  strong_retain %42 : $Builtin.BridgeObject       // id: %43
  %44 = struct $_StringObject (%41 : $UInt64, %42 : $Builtin.BridgeObject) // user: %45
  %45 = struct $_StringGuts (%44 : $_StringObject) // user: %46
  %46 = struct $String (%45 : $_StringGuts)       // user: %56
  %47 = struct_extract %37 : $Record, #Record.last // user: %48
  %48 = struct_extract %47 : $String, #String._guts // user: %49
  %49 = struct_extract %48 : $_StringGuts, #_StringGuts._object // users: %51, %50
  %50 = struct_extract %49 : $_StringObject, #_StringObject._countAndFlagsBits // user: %53
  %51 = struct_extract %49 : $_StringObject, #_StringObject._object // users: %53, %52
  strong_retain %51 : $Builtin.BridgeObject       // id: %52
  %53 = struct $_StringObject (%50 : $UInt64, %51 : $Builtin.BridgeObject) // user: %54
  %54 = struct $_StringGuts (%53 : $_StringObject) // user: %55
  %55 = struct $String (%54 : $_StringGuts)       // user: %56
  %56 = struct $Record (%46 : $String, %55 : $String) // user: %57
  store %56 to %36 : $*Record                     // id: %57
```

I think something isn't re-running as appropriate. I am going to try to just update the pass to do a little bit of simplification as it goes.

**Comment by gottesmm** [comment]
Took a deeper look here. The problem that we are running into here is that lower aggregate instrs now due to OSSA funnels values through reformed structs.

We run SILCombine right after we do this which shrinks loads when it canonicalizes them (load (struct_extract)) -> (struct_elt_addr (load)). This causes problems for later passes. Instead, we should first eliminate a bunch of these redundant extracts first before canonicalizing by adding a run of high level CSE. I would argue that we should /always/ run CSE before we re-canonicalize like this. I am going to try adding a run right after lower aggregate instrs. That fixes the regression in Phonebook.

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
There is one small remaining difference in Suffix.o, but it is due to an inlining decision change due to CSE happening where I put it.

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke test compiler performance

**Comment by gottesmm** [comment]
@swift-ci smoke compiler performance

**Comment by gottesmm** [comment]
@swift-ci Please smoke test compiler performance

**Comment by gottesmm** [comment]
Seems like putting CSE here makes a bigger change than I want to make here. I thought about this a little bit and I realized that:

1. When updating to OSSA, we are not going to split things up, so actually I think this is a problem we may not have to solve.
2. With that in mind, I can pre-optimize how we create these when we are not in OSSA.

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
I implemented what I said above. I need to fix one test but I want to see the code-size impact.

**Comment by gottesmm** [comment]
Exactly what I was looking for!

**Comment by gottesmm** [comment]
Fixed the test.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
@swift-ci clean test linux platform

**Pull request merged**